File: yt-agent-azure.py
"""Run this model in Python

> pip install azure-ai-inference
"""
import os
from azure.ai.inference import ChatCompletionsClient
from azure.ai.inference.models import AssistantMessage, SystemMessage, UserMessage, ToolMessage
from azure.ai.inference.models import ImageContentItem, ImageUrl, TextContentItem
from azure.core.credentials import AzureKeyCredential

# To authenticate with the model you will need to generate a personal access token (PAT) in your GitHub settings.
# Create your PAT token by following instructions here: https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens
client = ChatCompletionsClient(
    endpoint = "https://models.github.ai/inference",
    credential = AzureKeyCredential(os.environ["GITHUB_TOKEN"]),
)

messages = [
    UserMessage(content = [
        TextContentItem(text = "\n\nDownload all content from a YouTube channel, process into multiple formats, transcribe audio, and queue metadata generation via Ollama.\n\n# Steps\n1. **Download Source Content**:\n   - Fetch all videos, playlists, and Shorts from the specified YouTube channel.\n   - Save as MP4 files in the highest available quality.\n   - Use `yt-dlp` for downloading. Playlists ID for folder names if available, AND video ID for filenames.\n   - Use `ffmpeg` for any necessary format conversions.\n   - Use `whisper-cli` for audio transcription.\n\n2. **Format Conversion**:\n   - Create two additional copies of each MP4 file:\n     - **MP3**: Audio-only version.\n     - **16-bit WAV**: Uncompressed audio for transcription.\n\n3. **Transcription**:\n   - Process each WAV file using `whisper-cli` to generate a text transcript.\n   - Save transcripts as `.txt` files with matching filenames.\n\n4. **Metadata Queue**:\n   - Add all transcript files to a processing queue.\n   - Use Ollama to analyze transcripts and suggest improved titles/descriptions.\n\n5. Uploads a mirror storage to Cloudflare R2 Bucket.\n\n---\n\n# Output Format\nReturn a structured JSON object for each processed video, containing:\n```json\n{\n  \"source_video\": \"[YouTube URL]\",\n  \"downloaded_files\": {\n    \"mp4\": \"[filename].mp4\",\n    \"mp3\": \"[filename].mp3\",\n    \"wav\": \"[filename].wav\"\n  },\n  \"transcription\": {\n    \"status\": \"success/error\",\n    \"transcript_file\": \"[filename].txt\",\n    \"error_details\": \"[if applicable]\"\n  },\n  \"ollama_metadata_queue\": {\n    \"status\": \"queued/processed\",\n    \"suggested_title\": \"[Ollama-generated title]\",\n    \"suggested_description\": \"[Ollama-generated description]\"\n  }\n}\n```\n\n# Example\n**Input Channel**: `https://youtube.com/@example-channel`\n\n**Output**:\n```json\n{\n  \"source_video\": \"https://youtube.com/watch?v=abc123\",\n  \"downloaded_files\": {\n    \"mp4\": \"lecture_10.mp4\",\n    \"mp3\": \"lecture_10.mp3\",\n    \"wav\": \"lecture_10.wav\"\n  },\n  \"transcription\": {\n    \"status\": \"success\",\n    \"transcript_file\": \"lecture_10.txt\",\n    \"error_details\": null\n  },\n  \"ollama_metadata_queue\": {\n    \"status\": \"queued\",\n    \"suggested_title\": null,\n    \"suggested_description\": null\n  }\n}\n```\n\n# Notes\n- **Dependencies**: Ensure `yt-dlp`, `ffmpeg`, `whisper-cli`, and Ollama are installed and configured.\n- **Naming Convention**: Use consistent filenames across formats (e.g., `[video_id].mp4`, `[video_id].txt`).\n- **Placeholders**: Replace `[YouTube URL]`, `[filename]`, and Ollama-generated fields with actual values during execution."),
    ]),
]

tools = []

while True:
    response = client.complete(
        messages = messages,
        model = "deepseek/DeepSeek-R1",
        tools = tools,
    )

    if response.choices[0].message.tool_calls:
        print(response.choices[0].message.tool_calls)
        messages.append(response.choices[0].message)
        for tool_call in response.choices[0].message.tool_calls:
            messages.append(ToolMessage(
                content=locals()[tool_call.function.name](),
                tool_call_id=tool_call.id,
            ))
    else:
        print(f"[Model Response] {response.choices[0].message.content}")
        break

File: yt-agent-azure.php
Thinking...
Here's a thinking process that leads to the suggested PHP solution:

1.  **Analyze the Request:**
    *   **Input:** A Python script using the `azure-ai-inference` SDK to interact with a GitHub-hosted model (`deepseek/DeepSeek-R1`) via a Chat Completions API. It involves message handling, tool calling (though tools are initialized empty in the snippet), and output formatting.
    *   **Goal:** Convert this Python code into PHP.
    *   **Further Requirement:** "Revise and improve" the code. This means not just a direct port, but making it better structured, handling errors properly, following PHP best practices, and ensuring the logic makes sense (e.g., the Python code has a `while True` loop but defines `tools = []`, which might be incomplete).
    *   **Context:** The script attempts to handle a complex AI workflow (YouTube download -> Transcription -> Metadata generation). *Correction:* Looking closely at the Python code, the `messages` array contains a *prompt* that describes a complex workflow (using `yt-dlp`, `ffmpeg`, `whisper-cli`, `Cloudflare R2`, `Ollama`). However, the Python SDK (`azure-ai-inference`) is used to *chat* with an LLM. The Python script *calls the LLM* to generate JSON responses, but it does *not* actually execute the shell commands (`yt-dlp`, `ffmpeg`, etc.) within the script itself. The LLM is expected to output the results or the script is supposed to orchestrate external tools based on the LLM's output.
    *   *Critical Observation on the Python Code:* The Python code defines a `ChatCompletionsClient` pointing to `https://models.github.ai/inference`. This is a specific endpoint (GitHub Models). The code defines `messages` with a very long, multi-step instruction prompt. The loop handles tool calls (`tool_calls`). However, `tools = []` is defined. Usually, tools are defined separately. The Python code tries to call `locals()[tool_call.function.name]()`, which relies on functions defined outside this scope (in `locals()`), which won't work in the way intended for a generic AI orchestrator unless specific global functions exist for the tool execution.
    *   *Constraint:* The user wants PHP. PHP generally does not have native async libraries for Azure/GH AI Inference out of the box in the same way Python does. We need to look for PHP SDKs.
    *   *Availability Check:* Does a PHP SDK exist for `azure-ai-inference`? No, Azure's specific AI Inference SDK is Python-centric. GitHub Models might have its own PHP client or rely on raw HTTP requests (since the endpoint is `https://models.github.ai`). Given the lack of an official PHP SDK for `azure-ai-inference` or the specific GitHub Models endpoint, the *most robust* and "improved" way to do this in PHP is to use the **official Azure SDK for PHP** (which often supports AI Inference) or use **cURL/Guzzle** to call the API directly.
    *   *Better Approach:* For a generic AI inference task (like this GitHub Models endpoint), using **GuzzleHttp** to call the REST API directly is often more portable and "improved" than forcing a specific SDK if one isn't standard. However, there is an `azosdk` (Azure SDK) in PHP for Azure AI. Let's check if `azure-ai-inference` has PHP support. It does not seem to have a dedicated public SDK matching Python's API signature.
    *   *Decision:* The best "improved" PHP version will use **GuzzleHttp** to hit the OpenAI/Chat Completions compatible endpoint directly (most LLM APIs support OpenAI format). It will handle the JSON parsing, token management, and logic flow better than a raw `while True` loop (which can hang). I need to handle the "Tool Calling" aspect, which is complex. For the "improved" part, I should simplify the workflow logic because the Python script provided is actually an *orchestration* script where the LLM *thinks* it's doing work, but the script just waits for text.
    *   *Security:* Hardcoded tokens are bad. Use environment variables (like the Python script).
    *   *Structure:* Break down the prompt handling and the loop into a class structure (Service Pattern).

2.  **Refining the Logic (The "Improve" part):**
    *   **Python Script Analysis:** The Python script asks the LLM to perform tasks (`yt-dlp`, etc.) *and* upload to Cloudflare R2. The LLM cannot do this. The LLM generates text. The PHP script should likely generate the *response* based on the prompt, or it needs a local agent to actually execute the tools. Since I am converting the Python code *as is*, I must assume the PHP script is meant to act as the *client* that talks to the API, not the agent executing the `yt-dlp` commands.
    *   *Correction:* The Python code is a *client* that waits for the model. The prompt inside the Python code tells the model "Do this work". The model outputs a JSON. The client *needs* to parse that JSON and potentially execute the logic described in the prompt (or just collect the JSON).
    *   *Major Issue:* The Python code uses `locals()[tool_call.function.name]()` which implies it expects the PHP script to *actually* execute system commands or API calls defined in PHP functions to fulfill the tools requested by the LLM. In a standard "Chat" context, we usually don't expose `locals` tools like that.
    *   *Revised Plan:* I will implement a PHP class that uses Guzzle to interact with the API. I will handle the Chat Loop (Messages -> Model -> New Messages). I will clean up the prompt injection to be safer. I will replace the raw `locals` tool execution with a structured approach (simulating tool execution if we had them, or just handling text completion since `tools = []`). Since the Python code defines `tools = []`, the loop essentially waits for the model to stop generating tool calls.
    *   *Wait, actually:* The prompt asks to convert this specific logic. If I remove the tool logic, it's not a conversion. If I keep it, PHP doesn't have `locals()[...]` magic. I need to simulate tool execution or use a proper Tool Definition structure.
    *   *Best Practice:* The improved version should separate the *LLM Client* from the *Tool Logic*. The Python code is a "Local Agent" loop. In PHP, this means a `Loop` class that fetches, processes, and appends messages.
    *   *Endpoint:* `https://models.github.ai/inference` requires OpenAI-compatible Chat Completions format.

3.  **Step-by-Step Conversion Plan:**
    *   **Dependencies:** `guzzlehttp/guzzle`.
    *   **Client:** Use `GuzzleHttp\Client` to send `POST` requests.
    *   **Authentication:** `AzureKeyCredential` becomes an Authorization Header or Query Param depending on API structure. GitHub Models usually accepts Bearer tokens.
    *   **Message Structure:** Map Python `UserMessage`, `SystemMessage` etc., to standard JSON structures the API expects (`role`, `content`).
    *   **Loop:** Implement the `while` loop properly with a break condition (max iterations or stop condition).
    *   **Safety:** Use `.env` for secrets. Add error handling.
    *   **Tools:** Since `tools = []` in the original, I will implement the logic to handle potential tool calls if defined, but currently empty.
    *   **Code Quality:** Use Classes, Type Hinting (PHP 8+), Interfaces (optional but good), Logging.

4.  **Drafting the Code (Mental or Scratchpad):**
    *   Need `GuzzleHttp`.
    *   Need to serialize the complex nested prompt.
    *   Need to handle the `ToolMessage` equivalent (which is just an object in JSON).
    *   The `content` in the Python script is a list of dicts. In the LLM JSON, this is usually `{"role": "user", "content": [{"type": "text", "text": "..."}]}`.
    *   *Wait, GitHub Models API:* Does it expect Azure SDK specific types or OpenAI standard? The Python code uses `azure-ai-inference`. The endpoint is `models.github.ai`. This usually follows the OpenAI format but with GitHub Auth. The `azure-ai-inference` Python library handles this.
    *   *PHP Equivalent:* There is no official `azure-ai-inference` PHP SDK yet. We must use **cURL/Guzzle** and send raw JSON to the OpenAI-compatible endpoint.
    *   *Prompt Parsing:* The Python script has a complex `content` array with `TextContentItem`. I should clean this up into a standard string or a structured JSON for the API.
    *   *Loop Logic:* The Python code appends `ToolMessage` if tools are used. I need to replicate this logic using an associative array for tool responses.

5.  **Improving the Code:**
    *   **Security:** Sanitize inputs, use Environment Variables.
    *   **Error Handling:** Add `try-catch` blocks.
    *   **Logging:** Don't just `print`, use a logger (Psr-3 or basic).
    *   **Concurrency:** Ensure the loop terminates.
    *   **Architecture:** Create a class `YouTubeMetadataAgent`.

6.  **Refining the API Interaction:**
    *   The Python code: `response = client.complete(...)`.
    *   The equivalent in PHP: Guzzle `->post($url, ['json' => $payload])`.
    *   The Python code uses `AssistantMessage`, `SystemMessage` classes. I will construct the JSON manually to ensure compatibility with the `models.github.ai` endpoint (which typically expects OpenAI format).
    *   *Crucial Fix:* The Python code sends `UserMessage` with complex content. The GitHub API expects `{"role": "user", "content": [{"type": "text", "text": "..."}]}`. I will construct this JSON array explicitly.
    *   *Tool Calls:* The Python code checks `tool_calls`. I need to handle `tool_calls` in the response JSON if present.
    *   *Tool Execution:* In the improved version, I will abstract the tool execution. Instead of `locals[...]`, I will use a `ToolExecutor` method.

7.  **Final Code Structure:**
    *   `AgentConfig`: Holds credentials and settings.
    *   `AgentClient`: Handles the HTTP requests (Guzzle).
    *   `AgentLoop`: Handles the conversation history and tool execution.
    *   Main script calls the Agent.

8.  **Wait, one specific detail:** The user wants me to *transform* the provided Python code. The Python code is actually a *prompt* that asks the AI to download YouTube videos, etc. The AI *cannot* actually download files from YouTube. The Python script is an *orchestration script*. If I translate it to PHP, it becomes a PHP script that sends a request to an API. The API *might* have tool use capability, or the user expects the PHP script to handle the file downloads based on LLM instructions.
    *   *Reality Check:* You cannot ask an LLM hosted at `github.ai/inference` to download files directly via HTTP without it calling specific tools. The Python script assumes the LLM *will* have function calling capabilities to trigger external tools.
    *   *Simplification:* For the PHP conversion to be functional, I will treat the prompt as a "Chat Completions" request. I will implement the tool-calling logic as if the API supports it (standard OpenAI tool calling format). Since I cannot implement the `yt-dlp` shell commands in a generic LLM client without an external agent framework, I will create a *simulation* layer for tools or focus on the Chat Completions aspect.
    *   *Decision:* I will implement a robust PHP SDK wrapper using Guzzle that mimics the Python SDK's behavior. I will clean up the message structure. I will make the tool handling explicit (define tools, execute them).
    *   *Refinement:* Since the Python script sets `tools = []`, the model will *not* call tools unless the system prompt tells it to. The prompt provided in the Python code is a single User Message.
    *   *Safety:* I need to make sure the PHP code doesn't expose the token.

    *(Self-Correction during drafting)*: The Python script uses `azure-ai-inference`. That library handles specific authentication (Azure Key). GitHub Models API usually expects a Bearer Token. I will use the `Authorization: Bearer $TOKEN` header.

    Let's write the code.

    **Key Improvements to add:**
    1.  **Dependency Management:** Require `guzzlehttp/guzzle` and `ramsey/uuid`.
    2.  **Message Class:** Create a simple PHP class or array structure for messages to keep it clean, but raw JSON is often easier for direct API calls. I'll use a `stdClass` approach for JSON mapping.
    3.  **Tool Handling:** Instead of `locals()`, use a `handleToolCalls` function that parses the response and updates state.
    4.  **Error Handling:** Graceful fallback on network errors or rate limits.
    5.  **Formatting:** Use PHPDoc blocks, proper indentation.

    *(Wait, one more check)*: The Python code includes a prompt asking to download videos from YouTube. This is a sensitive prompt (scraping YouTube). I should warn the user or ensure the code focuses on the *LLM Client* aspect, not the *botnet/scraping* aspect. I will keep the prompt text but ensure the code is secure.
    *(Actually)*: The prompt asks to "Transform this into php, then revise and improve". The revision should focus on code quality and feasibility.
    *(Important)*: PHP doesn't have